chalice/app.py

```diff
@@ -138,24 +139,74 @@ def maybe_fake_504() -> typing.Optional[chalice.Response]:
+            status_class = str(status_code)[0] + 'xx'
+            try:
+                cloudwatch.put_metric_data(
+                    Namespace='DSS',
+                    MetricData=[
+                        dict(
+                            MetricName='HttpRequests by StatusClass',
+                            Dimensions=[
+                                dict(Name='StatusClass', Value=status_class),
+                            ],
+                            Value=1.0,
+                            Unit='Count',
+                            StorageResolution=60
+                        ),
+                        dict(
+                            MetricName='HttpRequests by Path, Method, StatusCode',
+                            Dimensions=[
+                                dict(Name='Path', Value=path_pattern),
+                                dict(Name='Method', Value=method),
+                                dict(Name='StatusCode', Value=str(status_code))
+                            ],
+                            Value=1.0,
+                            Unit='Count',
+                            StorageResolution=60
+                        ),
+                    ]
+                )
+            except Exception as e:
+                app.log.exception('Metrics post failed!', e)
+
+        status_code = None
+        try:
+            with flask_app.test_request_context(
+                    path=path,
+                    base_url="https://{}".format(app.current_request.headers["host"]),
+                    query_string=app.current_request.query_params,
+                    method=app.current_request.method,
+                    headers=list(app.current_request.headers.items()),
+                    data=req_body,
+                    environ_base=app.current_request.stage_vars):
+                with nestedcontext.bind(
+                        time_left=lambda: (
+                            (app.lambda_context.get_remaining_time_in_millis() / 1000) -
+                            EXECUTION_TERMINATION_THRESHOLD_SECONDS),
+                        skip_on_conflicts=True):
+                    flask_res = flask_app.full_dispatch_request()
+                    status_code = flask_res._status_code
+        finally:
+            app.log.info(
+                "[dispatch] \"%s %s\" %s%s",
+                method,
+                path,
+                str(status_code),
+                ' ' + str(query_params) if query_params is not None else '',
+            )
+
         # API Gateway/Cloudfront adds a duplicate Content-Length with a different value (not sure why)
+        res_headers = dict(flask_res.headers)
         res_headers.pop("Content-Length", None)
-        return chalice.Response(status_code=flask_res._status_code,
+
+        # observability logic is deferred to a different thread so as not to delay response
+        if metrics_executor._work_queue.qsize() >= MAX_EXECUTOR_WORKERS:
```

Comment:
What do you think about making the threshold 4 and the max pool size 8? And then also log the current size.

Independently, I would log as early as possible *before* adding tasks.

And I would log the size of both pools.

Reply:
Good suggestions, ✅✅✅